src/ruby/puppet-server-lib/puppet/server/http_client.rb

```diff
@@ -35,8 +34,8 @@ def initialize(server, port, options = {})
 
     @server = server
     @port = port
-    @use_ssl = options[:use_ssl]
-    @protocol = @use_ssl ? "https" : "http"
+    self.class.use_ssl = options[:use_ssl]
```

Comment:
The current logic derives the `hostcert`, `hostprivkey`, and `localcacert` from the global Puppet settings, right -- i.e., not passed in on a per request basis to a `Puppet::Server::HttpClient` instance?  I believe those settings will always have a non-empty value, with a default imposed by core Ruby Puppet if the user hasn't overridden it in their puppet.conf file.

I had been thinking that clients creating `Puppet::Server::HttpClient` instances just specify whether or not SSL needs to be used and not what certs and keys to use.  I'm not suggesting that this change for this PR.

Maybe the only concerning case would be for the case that Puppet Server were being configured to run via HTTP -- the external SSL termination case.  If it is possible for Puppet to come to ready with the values for `hostcert`, `hostprivkey`, and `localcacert` pointing to files that don't actually exist then trying to configure the SSL context for HTTP client calls would be problematic.  If that were the case, maybe the client logic could be made more robust here by checking that all of the files exist before trying to configure the SSL context and, if they do not, outputting a log warning and skipping the SSL context configuration.  If a client were to later try to make a request with a "use_ssl" of true, Puppet Server could fail the request with an error indicating that the context could not be configured because certs/keys were missing.  I'm thinking this would have been an issue with this code prior to all of the work done to optimize the number of clients being created, though.  If this isn't a valid use case, adding in this extra robustness wouldn't seem to be necessary.

@cprice404 - what do you think?